## Fundamentos de la Programación (CDIA)
### Convocatoria ordinaria, junio de 2025

**Nombre:** Álvaro 

**Apellidos:** Gallo Alonso

### Enunciado: Sistema de Encuestas de Satisfacción
Se desea construir un sistema para analizar encuestas de satisfacción de clientes, utilizando programación orientada a objetos y únicamente las librerías estándar de Python vistas a lo largo del curso. Para ello se cuenta con los siguientes enumerados ya definidos que deben ser utilizados por la implementación (no deben ser modificados):


In [1]:
from enum import Enum

# Nos permite definir el tipo de comentario realizado por el usuario
class TipoComentario(Enum):
    POSITIVO = "positivo"
    NEGATIVO = "negativo"
    NEUTRO = "neutro"
    INAPROPIADO = "inapropiado"
    SIN_CLASIFICAR = "sin clasificar"
    SIN_COMENTARIO = "sin comentario"

    def __repr__(self) -> str:
        return f'{self}'

# Define los campos recogidos en la encuesta
class CampoEncuesta(Enum):
    ATENCION = "atención"
    CALIDAD = "calidad"
    PRECIO = "precio"
    TIEMPO_ENTREGA = "tiempo de entrega"
    FACILIDAD_USO = "facilidad de uso"

    def __repr__(self) -> str:
        return f'{self}'


#### Clases a implementar

1. Clase __ComentarioClasificado__ (1 punto)
 - Atributos/Propiedades:
   - __tipo__:_TipoComentario_, recoge el tipo de comentario realizado.
   - __texto__:_str_, recoge la cadena de texto del comentario realizado.
 - Validaciones en el constructor:
   - La llamada al constructor sin parámetros deberá crear un comentario por defecto con:
     -  __tipo__ → _TipoComentario.SIN_COMENTARIO_ y 
     -  __texto__ → _vacío_.
   - También admite una llamada con los parámetros __tipo__:_TipoComentario_ y __texto__:_str_. 
     - Si el __tipo__ es _SIN_COMENTARIO_ de proporcionarse el parámetro __texto__ éste se ignorará, siendo su valor la cadena vacía.
     - El __texto__ es obligatorio si el __tipo__ es distinto de _SIN_COMENTARIO_ (no puede ser la cadena vacía).
   - Se deben lanzar excepciones apropiadas (_ValueError_, _TypeError_, _AttributeError_, _KeyError_, etc.) si los parámetros proporcionados no son del tipo adecuado.
 - Métodos:
   - Reescribir el método **\_\_str\_\_** para que genere una salida como la del ejemplo siguiente:
   <pre style='font-size:12px; color:blue'>
            [POSITIVO] Muy buen servicio.
   </pre>
   - Reescribir el método **\_\_repr\_\_** adecuadamente para que su salida pueda ser evaluada y genere el correspondiente objeto de tipo _ComentarioClasificado_.
 - Información adicional:
   - Los objetos instanciados deben ser inmutables (propiedades de solo lectura).
<br/><br/>
2. Clase __Encuesta__ (4 puntos)
 - Atributos/Propiedades:
   - __cliente_id__:_str_, cadena con 4 letras seguidas de 4 números (ej. "ABCD1234") que identifica al cliente.
   - __respuestas__:_dict[CampoEncuesta, int]_, diccionario cuyas claves son las presentes en el tipo _CampoEncuesta_ y cuyos valores son enteros entre 0 y 5. 
   - __comentario__:_ComentarioClasificado_, comentario clasificado.
 - Validaciones en el constructor:
   - El constructor recibe como parámetros:
     - __cliente_id__:_str_, debe cumplir el patrón indicado (usando expresión regular),
     - un __diccionario__:_dict[CampoEncuesta, int]_ con al menos una entrada (a los campos de la encuesta no valorados se le asigna el valor 0) y valores enteros entre 0 y 5. 
     - [_opcional_] __comentario__:_ComentarioClasificado_. En caso de no proporcionarse se incluirá un comentario por defecto.
   - Si no se cumplen las validaciones, se deben lanzar excepciones apropiadas (_ValueError_, _TypeError_, _AttributeError_, _KeyError_, etc.). En el caso de campos de la encuesta no válidos, éstos se deberán mostrar en el texto de la excepción.
 - Métodos:
   - __promedio()__ -> float: devuelve el promedio de los valores registrados en la encuesta.
   - Reescribir el método **\_\_str\_\_** para que genere una salida como la del ejemplo siguiente:
   <pre style='font-size:12px; color:blue'>
          Encuesta de cliente 'ABCD1234':
            Respuestas → atención: 4, calidad: 5, precio: 3, tiempo de entrega: 0, facilidad de uso: 5
            Comentario → [POSITIVO] Excelente atención.
   </pre>
   - Reescribir el método **\_\_repr\_\_** adecuadamente para que su salida pueda ser evaluada y genere el correspondiente objeto del tipo _Encuesta_.
 - Información adicional:
   - Los objetos instanciados deben ser inmutables (propiedades de solo lectura).
<br/><br/>
3. Clase __AnalizadorEncuestas__ (5 puntos)
 - Atributos/Propiedades:
   - __encuestas__:_list[Encuesta]_. Lista de encuestas.
 - Validaciones en el constructor:
   - Constructor por defecto, sin parámetros.
 - Métodos:
   - __agregar_encuesta(encuesta: Encuesta)__ -> None: añade una encuesta, verificando su tipo. Debe lanzar una excepción apropiada (_ValueError_, _TypeError_, _AttributeError_, _KeyError_, etc.) si el parámetro no es del tipo adecuado.
   - __promedio_por_campo()__ -> dict[CampoEncuesta, float]: valoración promedio por campo.
   - __promedio_general()__ -> float: promedio total de todas las encuestas.
   - __resumen_votos()__ -> dict[CampoEncuesta, list[int]]: devuelve una lista de los valores registrados por campo.
   - __resumen_comentarios()__ -> dict[TipoComentario, int]: número de encuestas por tipo de comentario.
   - Reescribir el método **\_\_str\_\_** para que genere una salida como la del ejemplo siguiente:
      <pre style='font-size:12px; color:blue'>
          --- Encuesta #1 ---
          Encuesta de cliente 'ABCD1234':
            Respuestas → atención: 5, calidad: 4, precio: 3, tiempo de entrega: 4, facilidad de uso: 5
            Comentario → [POSITIVO] Muy bien
      </pre>
      <pre style='font-size:12px; color:blue'>
          --- Encuesta #2 ---
          Encuesta de cliente 'WXYZ5678':
            Respuestas → atención: 3, calidad: 2, precio: 4, tiempo de entrega: 3, facilidad de uso: 0
            Comentario → [NEGATIVO] Lento y caro
      </pre>
 - Información adicional:
   - La propiedad __encuestas__ es de solo lectura, sólo se puede modificar su contenido haciendo uso del método __agregar_encuesta__.
<br/><br/>
#### Evaluación

Se valorará:
 - Correcta aplicación de programación orientada a objetos.
 - Encapsulamiento, validación y uso correcto de excepciones.
 - Uso adecuado de enumeraciones, estructuras de datos y funciones.
 - Modularidad, claridad y organización del código.
 - Las versiones más óptimas del código.

In [10]:
# Clase ComentarioClasificado
class ComentarioClasificado:
    __slots__ = ("_tipo", "_texto")

    def __init__(self, tipo=None, texto=None):
        if tipo is None and texto is None:
            self._tipo = TipoComentario.SIN_COMENTARIO
            self._texto = ""
        elif tipo is not None:
            if not isinstance(tipo, TipoComentario):
                raise TypeError("El tipo debe ser una instancia de TipoComentario.")
            if tipo == TipoComentario.SIN_COMENTARIO:
                self._tipo = tipo
                self._texto = ""
            else:
                if texto is None:
                    raise ValueError("Debe proporcionar un texto si el tipo no es SIN_COMENTARIO.")
                if not isinstance(texto, str):
                    raise TypeError("El texto debe ser una cadena de caracteres.")
                if texto.strip() == "":
                    raise ValueError("Debe proporcionar un texto si el tipo no es SIN_COMENTARIO.")
                self._tipo = tipo
                self._texto = texto
        else:
            raise ValueError("Debe proporcionar el tipo si proporciona texto.")

    @property
    def tipo(self):
        return self._tipo

    @property
    def texto(self):
        return self._texto

    def __str__(self):
        return f"[{self._tipo.name}] {self._texto}".rstrip()

    def __repr__(self):
        if self._tipo == TipoComentario.SIN_COMENTARIO:
            return "ComentarioClasificado()"
        return f"ComentarioClasificado(TipoComentario.{self._tipo.name}, {self._texto!r})"

Se propociona el siguiente código, que no se debe modificar, para la comprobación básica del funcionamiento de la clase. Superar las pruebas no significa que el código sea correcto:

In [3]:
def probar_comentario_clasificado():

    print("=== PRUEBAS: ComentarioClasificado ===\n")

    print("1. Creación por defecto:")
    try:
        c = ComentarioClasificado()
        print(f"  ✔️ Comentario por defecto creado correctamente: {c}\n")  
    except Exception as e:
        print(f"  ❌ Error inesperado: {e}\n") 

    print("2. Creación con parámetros válidos:")
    try:
        c = ComentarioClasificado(TipoComentario.POSITIVO, "Muy buen servicio")
        print(f"  ✔️ Comentario con parámetros creado correctamente: {c}\n")
    except Exception as e:
        print(f"  ❌ Error inesperado: {e}\n")

    print("3. Creación con parámetros no válidos:")
    try:
        c = ComentarioClasificado(TipoComentario.POSITIVO)
        print("  ❌ No se lanzó excepción esperada.")
    except Exception as e:
        print(f"  ✔️ Excepción capturada correctamente: {e}")
    try:
        c = ComentarioClasificado(TipoComentario.SIN_COMENTARIO, "comentario")
        if c.texto == "":
            print(f"  ✔️ Comentario creado correctamente: {c}\n") 
        else: raise ValueError("El tipo SIN_COMENTARIO no puede tener texto asociado.") 
    except Exception as e:
        print(f"  ❌ Comentario con parámetros erróneos: {e}\n")

    print("4. Tipo inválido (cadena en lugar de enumerado):")
    try:
        ComentarioClasificado("positivo", "Comentario no válido")
        print("  ❌ No se lanzó excepción esperada.\n")
    except TypeError as e:
        print(f"  ✔️ Excepción capturada correctamente: {e}\n")

    print("5. Texto inválido (entero en lugar de cadena):")
    try:
        ComentarioClasificado(TipoComentario.NEUTRO, 12345)
        print("  ❌ No se lanzó excepción esperada.\n")
    except TypeError as e:
        print(f"  ✔️ Excepción capturada correctamente: {e}\n")

    print("6. Acceso a propiedades:")
    try:
        c = ComentarioClasificado(TipoComentario.NEGATIVO, "Pésima atención")
        tipo = c.tipo
        texto = c.texto
        print("  ✔️ Acceso a propiedades correcto.")
        print(f"    - Tipo: {tipo}")
        print(f"    - Texto: {texto}\n")
    except Exception as e:
        print(f"  ❌ Error al acceder a propiedades: {e}\n")

    print("7. Intento de modificación de propiedades:")
    try:
        c.tipo = TipoComentario.POSITIVO
        print("  ❌ No se lanzó excepción al modificar la propiedad 'tipo'.")
    except AttributeError as e:
        print("  ✔️ No se puede modificar la propiedad 'tipo'.")
    try:
        c.texto = "Nuevo comentario"
        print("  ❌ No se lanzó excepción al modificar la propiedad 'texto'.\n")
    except AttributeError as e:
        print("  ✔️ No se puede modificar la propiedad 'texto'.\n")

    print("8. Reescritura de __str__:")
    try:
        c = ComentarioClasificado(TipoComentario.POSITIVO, "Muy buen servicio")
        if str(c) == "[POSITIVO] Muy buen servicio":
            print("  ✔️ Método __str__ correctamente generado.\n")
        else:
            print("  ❌ Método __str__ incorrecto.\n")
    except:
        print("  ❌ Método __str__ incorrecto.\n")

    print("9. Reescritura de __repr__:")
    try:
        c = ComentarioClasificado(TipoComentario.POSITIVO, "Muy buen servicio")
        if isinstance(eval(repr(c)), ComentarioClasificado):
            print("  ✔️ Método __repr__ correctamente generado.")
        else:
            print("  ❌ Método __repr__ incorrecto.")
    except:
        print("  ❌ Método __repr__ incorrecto.")

# Ejecutar la prueba
probar_comentario_clasificado()

=== PRUEBAS: ComentarioClasificado ===

1. Creación por defecto:
  ✔️ Comentario por defecto creado correctamente: [SIN_COMENTARIO]

2. Creación con parámetros válidos:
  ✔️ Comentario con parámetros creado correctamente: [POSITIVO] Muy buen servicio

3. Creación con parámetros no válidos:
  ✔️ Excepción capturada correctamente: Debe proporcionar un texto si el tipo no es SIN_COMENTARIO.
  ✔️ Comentario creado correctamente: [SIN_COMENTARIO]

4. Tipo inválido (cadena en lugar de enumerado):
  ✔️ Excepción capturada correctamente: El tipo debe ser una instancia de TipoComentario.

5. Texto inválido (entero en lugar de cadena):
  ✔️ Excepción capturada correctamente: El texto debe ser una cadena de caracteres.

6. Acceso a propiedades:
  ✔️ Acceso a propiedades correcto.
    - Tipo: TipoComentario.NEGATIVO
    - Texto: Pésima atención

7. Intento de modificación de propiedades:
  ✔️ No se puede modificar la propiedad 'tipo'.
  ✔️ No se puede modificar la propiedad 'texto'.

8. Reescritur

__Salida esperada__:
<pre style='font-size:12px;color:blue'>
=== PRUEBAS: ComentarioClasificado ===

1. Creación por defecto:
  ✔️ Comentario por defecto creado correctamente: [SIN_COMENTARIO] 

2. Creación con parámetros válidos:
  ✔️ Comentario con parámetros creado correctamente: [POSITIVO] Muy buen servicio

3. Creación con parámetros no válidos:
  ✔️ Excepción capturada correctamente: Debe proporcionar un texto si el tipo no es SIN_COMENTARIO.
  ✔️ Comentario creado correctamente: [SIN_COMENTARIO] 

4. Tipo inválido (cadena en lugar de enumerado):
  ✔️ Excepción capturada correctamente: El tipo debe ser una instancia de TipoComentario.

5. Texto inválido (entero en lugar de cadena):
  ✔️ Excepción capturada correctamente: El texto debe ser una cadena de caracteres.

6. Acceso a propiedades:
  ✔️ Acceso a propiedades correcto.
    - Tipo: TipoComentario.NEGATIVO
    - Texto: Pésima atención

7. Intento de modificación de propiedades:
  ✔️ No se puede modificar la propiedad 'tipo'.
  ✔️ No se puede modificar la propiedad 'texto'.

8. Reescritura de __str__:
  ✔️ Método __str__ correctamente generado.
  
9. Reescritura de __repr__:
  ✔️ Método __repr__ correctamente generado.
</pre>

In [11]:
# Clase Encuesta
import re

class Encuesta:
    __slots__ = ("_cliente_id", "_respuestas", "_comentario")

    def __init__(self, cliente_id, respuestas, comentario=None):
        # Validar cliente_id
        if not isinstance(cliente_id, str):
            raise TypeError("El identificador del cliente debe ser una cadena.")
        if not re.fullmatch(r"[A-Za-z]{4}\d{4}", cliente_id):
            raise ValueError("El identificador del cliente debe tener 4 letras seguidas de 4 dígitos.")
        self._cliente_id = cliente_id.upper()

        # Validar respuestas
        if not isinstance(respuestas, dict):
            raise TypeError("Las respuestas deben ser un diccionario.")
        if not respuestas:
            raise ValueError("Debe proporcionar al menos una respuesta.")

        claves_no_validas = [k for k in respuestas if not isinstance(k, CampoEncuesta)]
        if claves_no_validas:
            nombres = [str(k) for k in respuestas if not isinstance(k, CampoEncuesta)]
            raise TypeError(f"Las claves {nombres} no son del tipo CampoEncuesta.")

        for v in respuestas.values():
            if not (isinstance(v, int) and 0 <= v <= 5):
                raise ValueError("Todos los valores en las respuestas deben ser enteros entre 0 y 5.")

        # Completar los campos que faltan con 0
        respuestas_completas = {}
        for campo in CampoEncuesta:
            respuestas_completas[campo] = respuestas.get(campo, 0)
        self._respuestas = respuestas_completas

        # Validar comentario
        if comentario is None:
            comentario = ComentarioClasificado()
        if not isinstance(comentario, ComentarioClasificado):
            raise TypeError("El comentario debe ser de tipo ComentarioClasificado.")
        self._comentario = comentario

    @property
    def cliente_id(self):
        return self._cliente_id

    @property
    def respuestas(self):
        return self._respuestas.copy()

    @property
    def comentario(self):
        return self._comentario

    def promedio(self):
        valores = list(self._respuestas.values())
        return round(sum(valores) / len(valores), 2)

    def __str__(self):
        respuestas_str = ", ".join(
            f"{campo.value}: {self._respuestas[campo]}" for campo in CampoEncuesta
        )
        return (
            f"Encuesta de cliente '{self._cliente_id}':\n"
            f"  Respuestas → {respuestas_str}\n"
            f"  Comentario → {self._comentario}"
        )

    def __repr__(self):
        return (
            f"Encuesta({self._cliente_id!r}, "
            f"{{{', '.join(f'{k!r}: {v!r}' for k, v in self._respuestas.items())}}}, "
            f"{repr(self._comentario)})"
        )

Se propociona el siguiente código, que no se debe modificar, para la comprobación básica del funcionamiento de la clase. Superar las pruebas no significa que el código sea correcto:

In [12]:
def probar_encuesta():

    print("\n=== PRUEBAS: Encuesta ===\n")

    respuestas_validas = {
        CampoEncuesta.ATENCION: 4,
        CampoEncuesta.CALIDAD: 5,
        CampoEncuesta.PRECIO: 3,
        # CampoEncuesta.TIEMPO_ENTREGA: 4,
        CampoEncuesta.FACILIDAD_USO: 5
    }
    promedio_valido = 3.40

    comentario = ComentarioClasificado(TipoComentario.POSITIVO, "Excelente atención.")

    print("1. Creación correcta con comentario:")
    try:
        e1 = Encuesta("ABcd1234", respuestas_validas, comentario)
        e1.respuestas[CampoEncuesta.TIEMPO_ENTREGA]
        print("  ✔️ Encuesta creada correctamente.\n")
        print(e1,'\n')
    except Exception as e:
        print(f"  ❌ Error inesperado: {e}\n")

    print("2. Creación sin comentario (valor por defecto):")
    try:
        e1 = Encuesta("WXYZ9999", respuestas_validas)
        print("  ✔️ Encuesta sin comentario creada correctamente.\n")
        print(e1,'\n')
    except Exception as e:
        print(f"  ❌ Error inesperado: {e}\n")

    print("3. Cliente ID inválido:")
    try:
        Encuesta("1234ABCD", respuestas_validas, comentario)
        print("  ❌ No se lanzó excepción esperada por cliente_id inválido.\n")
    except ValueError as e:
        print(f"  ✔️ Excepción capturada correctamente: {e}\n")

    print("4. Clave no válida en respuestas:")
    try:
        respuestas_invalidas = {"diseño": 5, CampoEncuesta.ATENCION: 4, 'colores': 3}
        Encuesta("TEST1234", respuestas_invalidas, comentario)
        print("  ❌ No se lanzó excepción esperada por clave inválida.\n")
    except TypeError as e:
        print(f"  ✔️ Excepción capturada correctamente: {e}\n")

    print("5. Valor fuera de rango en respuestas:")
    try:
        respuestas_invalidas = dict(respuestas_validas)
        respuestas_invalidas[CampoEncuesta.PRECIO] = 8
        respuestas_invalidas[CampoEncuesta.FACILIDAD_USO] = 9
        Encuesta("TEST1234", respuestas_invalidas, comentario)
        print("  ❌ No se lanzó excepción esperada por valor fuera de rango.\n")
    except ValueError as e:
        print(f"  ✔️ Excepción capturada correctamente: {e}\n")

    print("6. Acceso a propiedades:")
    try:
        cli = e1.cliente_id
        com = e1.comentario.texto
        res = e1.respuestas
        print("  ✔️ Acceso correcto a propiedades.\n")
        print(f"   - ID: {cli}")
        print(f"   - Comentario: {com}")
        print(f"   - Respuestas: {res}\n")
    except Exception as e:
        print(f"  ❌ Error al acceder a propiedades: {e}\n")

    print("7. Intento de modificación de propiedades:")
    try:
        e1.cliente_id = "ZZZZ0000"
        print("  ❌ No se lanzó excepción al modificar 'cliente_id'.\n")
    except AttributeError as e:
        print(f"  ✔️ No se puede modificar la propiedad 'cliente_id'.")
    try:
        e1.respuestas = {}
        print("  ❌ No se lanzó excepción al modificar 'respuestas'.\n")
    except AttributeError as e:
        print(f"  ✔️ No se puede modificar la propiedad 'respuestas'.")
    try:
        e1.comentario = {}
        print("  ❌ No se lanzó excepción al modificar 'comentario'.\n")
    except AttributeError as e:
        print(f"  ✔️ No se puede modificar la propiedad 'comentario'.\n")

    print("8. Cálculo del promedio:")
    try:
        if (promedio_valido == e1.promedio()):
            print(f"  ✔️ Promedio calculado correctamente: {promedio_valido}\n")
        else: raise(Exception)
    except Exception as e:
        print(f"  ❌ Error al calcular promedio: {e}\n")

    print("9. Reescritura de __str__:")
    try:
        salid_esperada = '''Encuesta de cliente 'ABCD1234':
  Respuestas → atención: 4, calidad: 5, precio: 3, tiempo de entrega: 0, facilidad de uso: 5
  Comentario → [POSITIVO] Excelente atención.''' 
        e1 = Encuesta("ABcd1234", respuestas_validas, comentario)
        if str(e1) == salid_esperada:
            print("  ✔️ Método __str__ correctamente generado.\n")
        else:
            print("  ❌ Método __str__ incorrecto.\n")
    except:
        print("  ❌ Método __str__ incorrecto.\n")

    print("10. Reescritura de __repr__:")
    try:
        e1 = Encuesta("ABcd1234", respuestas_validas, comentario)
        if isinstance(eval(repr(e1)), Encuesta):
            print("  ✔️ Método __repr__ correctamente generado.")
        else:
            print("  ❌ Método __repr__ incorrecto.")
    except:
        print("  ❌ Método __repr__ incorrecto.")

# Ejecutar las pruebas
probar_encuesta()
import re

class Encuesta:
    __slots__ = ("_cliente_id", "_respuestas", "_comentario")

    def __init__(self, cliente_id, respuestas, comentario=None):
        # Validación cliente_id
        if not isinstance(cliente_id, str):
            raise TypeError("El identificador del cliente debe ser una cadena.")
        if not re.fullmatch(r"[A-Za-z]{4}\d{4}", cliente_id):
            raise ValueError("El identificador del cliente debe tener 4 letras seguidas de 4 dígitos.")
        self._cliente_id = cliente_id.upper()

        # Validación respuestas
        if not isinstance(respuestas, dict):
            raise TypeError("Las respuestas deben ser un diccionario.")
        if not respuestas:
            raise ValueError("Debe proporcionar al menos una respuesta.")

        claves_no_validas = [k for k in respuestas if not isinstance(k, CampoEncuesta)]
        if claves_no_validas:
            nombres = [str(k) for k in respuestas if not isinstance(k, CampoEncuesta)]
            raise TypeError(f"Las claves {nombres} no son del tipo CampoEncuesta.")

        for v in respuestas.values():
            if not (isinstance(v, int) and 0 <= v <= 5):
                raise ValueError("Todos los valores en las respuestas deben ser enteros entre 0 y 5.")

        # Completar los campos que faltan con 0
        respuestas_completas = {}
        for campo in CampoEncuesta:
            respuestas_completas[campo] = respuestas.get(campo, 0)
        self._respuestas = respuestas_completas

        # Validación comentario
        if comentario is None:
            comentario = ComentarioClasificado()
        if not isinstance(comentario, ComentarioClasificado):
            raise TypeError("El comentario debe ser de tipo ComentarioClasificado.")
        self._comentario = comentario

    @property
    def cliente_id(self):
        return self._cliente_id

    @property
    def respuestas(self):
        return self._respuestas.copy()

    @property
    def comentario(self):
        return self._comentario

    def promedio(self):
        valores = list(self._respuestas.values())
        return round(sum(valores) / len(valores), 2)

    def __str__(self):
        respuestas_str = ", ".join(
            f"{campo.value}: {self._respuestas[campo]}" for campo in CampoEncuesta
        )
        return (
            f"Encuesta de cliente '{self._cliente_id}':\n"
            f"  Respuestas → {respuestas_str}\n"
            f"  Comentario → {self._comentario}"
        )

    def __repr__(self):
        return (
            f"Encuesta({self._cliente_id!r}, "
            f"{{{', '.join(f'{k!r}: {v!r}' for k, v in self._respuestas.items())}}}, "
            f"{repr(self._comentario)})"
        )


=== PRUEBAS: Encuesta ===

1. Creación correcta con comentario:
  ✔️ Encuesta creada correctamente.

Encuesta de cliente 'ABCD1234':
  Respuestas → atención: 4, calidad: 5, precio: 3, tiempo de entrega: 0, facilidad de uso: 5
  Comentario → [POSITIVO] Excelente atención. 

2. Creación sin comentario (valor por defecto):
  ✔️ Encuesta sin comentario creada correctamente.

Encuesta de cliente 'WXYZ9999':
  Respuestas → atención: 4, calidad: 5, precio: 3, tiempo de entrega: 0, facilidad de uso: 5
  Comentario → [SIN_COMENTARIO] 

3. Cliente ID inválido:
  ✔️ Excepción capturada correctamente: El identificador del cliente debe tener 4 letras seguidas de 4 dígitos.

4. Clave no válida en respuestas:
  ✔️ Excepción capturada correctamente: Las claves ['diseño', 'colores'] no son del tipo CampoEncuesta.

5. Valor fuera de rango en respuestas:
  ✔️ Excepción capturada correctamente: Todos los valores en las respuestas deben ser enteros entre 0 y 5.

6. Acceso a propiedades:
  ✔️ Acceso correc

__Salida esperada__:
<pre style='font-size:12px;color:blue'>
=== PRUEBAS: Encuesta ===

1. Creación correcta con comentario:
  ✔️ Encuesta creada correctamente.

Encuesta de cliente 'ABCD1234':
  Respuestas → atención: 4, calidad: 5, precio: 3, tiempo de entrega: 0, facilidad de uso: 5
  Comentario → [POSITIVO] Excelente atención. 

2. Creación sin comentario (valor por defecto):
  ✔️ Encuesta sin comentario creada correctamente.

Encuesta de cliente 'WXYZ9999':
  Respuestas → atención: 4, calidad: 5, precio: 3, tiempo de entrega: 0, facilidad de uso: 5
  Comentario → [SIN_COMENTARIO]  

3. Cliente ID inválido:
  ✔️ Excepción capturada correctamente: El identificador del cliente debe tener 4 letras seguidas de 4 dígitos.

4. Clave no válida en respuestas:
  ✔️ Excepción capturada correctamente: Las claves ['diseño', 'colores'] no son del tipo CampoEncuesta.

5. Valor fuera de rango en respuestas:
  ✔️ Excepción capturada correctamente: Todos los valores en las respuestas deben ser enteros entre 0 y 5.

6. Acceso a propiedades:
  ✔️ Acceso correcto a propiedades.

   - ID: WXYZ9999
   - Comentario: 
   - Respuestas: {CampoEncuesta.ATENCION: 4, CampoEncuesta.CALIDAD: 5, CampoEncuesta.PRECIO: 3, CampoEncuesta.TIEMPO_ENTREGA: 0, CampoEncuesta.FACILIDAD_USO: 5}

7. Intento de modificación de propiedades:
  ✔️ No se puede modificar la propiedad 'cliente_id'.
  ✔️ No se puede modificar la propiedad 'respuestas'.
  ✔️ No se puede modificar la propiedad 'comentario'.

8. Cálculo del promedio:
  ✔️ Promedio calculado correctamente: 3.4

9. Reescritura de __str__:
  ✔️ Método __str__ correctamente generado.

10. Reescritura de __repr__:
  ✔️ Método __repr__ correctamente generado.
</pre>

In [13]:
# Clase AnalizadorEncuestas
class AnalizadorEncuestas:
    __slots__ = ("_encuestas",)

    def __init__(self):
        self._encuestas = []

    @property
    def encuestas(self):
        return list(self._encuestas)

    def agregar_encuesta(self, encuesta):
        if not isinstance(encuesta, Encuesta):
            raise TypeError("Solo se pueden agregar objetos de tipo Encuesta.")
        self._encuestas.append(encuesta)

    def promedio_por_campo(self):
        total = {campo: 0 for campo in CampoEncuesta}
        cuenta = {campo: 0 for campo in CampoEncuesta}
        for e in self._encuestas:
            for campo in CampoEncuesta:
                total[campo] += e.respuestas[campo]
                cuenta[campo] += 1
        return {campo: (round(total[campo] / cuenta[campo], 2) if cuenta[campo] > 0 else 0.0) for campo in CampoEncuesta}

    


        

Se propociona el siguiente código, que no se debe modificar, para la comprobación básica del funcionamiento de la clase. Superar las pruebas no significa que el código sea correcto:

In [14]:
def probar_analizador_encuestas():
    print("=== PRUEBAS: AnalizadorEncuestas ===\n")

    respuestas1 = {
        CampoEncuesta.ATENCION: 5,
        CampoEncuesta.CALIDAD: 4,
        CampoEncuesta.PRECIO: 3,
        CampoEncuesta.TIEMPO_ENTREGA: 4,
        CampoEncuesta.FACILIDAD_USO: 5
    }

    respuestas2 = {
        CampoEncuesta.ATENCION: 3,
        CampoEncuesta.CALIDAD: 2,
        CampoEncuesta.PRECIO: 4,
        CampoEncuesta.TIEMPO_ENTREGA: 3,
    }

    promedio_por_campo = 4.00, 3.00, 3.50, 3.50, 2.50
    promedio_general = 3.30

    c1 = ComentarioClasificado(TipoComentario.POSITIVO, "Muy bien")
    c2 = ComentarioClasificado(TipoComentario.NEGATIVO, "Lento y caro")

    e1 = Encuesta("ABCD1234", respuestas1, c1)
    e2 = Encuesta("WXYZ5678", respuestas2, c2)

    print("1. Creación de AnalizadorEncuestas vacío:")
    try:
        analizador = AnalizadorEncuestas()
        print(f"  ✔️ Analizador creado con {len(analizador.encuestas)} encuestas.\n")
        print('\t', analizador, '\n')
    except Exception as e:
        print(f"  ❌ Error al crear AnalizadorEncuestas: {e}\n")

    print("2. Agregar encuestas:")
    try:
        analizador.agregar_encuesta(e1)
        analizador.agregar_encuesta(e2)
        analizador.encuestas.append(e2)
        if len(analizador.encuestas) == 2:
            print(f"  ✔️ Se agregaron {len(analizador.encuestas)} encuestas.\n")
            print(analizador,'\n')
        else: raise Exception('Sólo se pueden actualizar mediante el método \'agregar_encuesta\'')
    except Exception as e:
        print(f"  ❌ Error al agregar encuestas: {e}\n")

    print("3. Promedio por campo:")
    try:
        if promedio_por_campo == tuple(analizador.promedio_por_campo().values()):
            print(f"  ✔️ Promedio por campo calculado correctamente: {promedio_por_campo}\n")
        else: raise(Exception)
    except Exception as e:
        print(f"  ❌ Error al calcular promedio por campo: {e}\n")

    print("4. Promedio general:")
    try:
        if (promedio_general == analizador.promedio_general()):
            print(f"  ✔️ Promedio general calculado correctamente: {promedio_general:.2f}\n")
        else: raise(Exception)
    except Exception as e:
        print(f"  ❌ Error al calcular promedio general: {e}\n")

    print("5. Resumen de votos:")
    try:
        salida_esperada = {CampoEncuesta.ATENCION: [5, 3], CampoEncuesta.CALIDAD: [4, 2], CampoEncuesta.PRECIO: [3, 4], 
                            CampoEncuesta.TIEMPO_ENTREGA: [4, 3], CampoEncuesta.FACILIDAD_USO: [5, 0]}
        resumen = analizador.resumen_votos()
        if salida_esperada == resumen:
            print("  ✔️ Resumen de votos generado correctamente.\n")
            for campo, votos in resumen.items():
                print(f"   {campo.name:>14}: {str(votos):<8}")
            print()
        else: raise(Exception)
    except Exception as e:
        print(f"  ❌ Error al generar resumen de votos: {e}\n")

    print("6. Resumen de comentarios:")
    try:
        salida_esperada = {TipoComentario.POSITIVO: 1, TipoComentario.NEGATIVO: 1, TipoComentario.NEUTRO: 0,       
                           TipoComentario.INAPROPIADO: 0, TipoComentario.SIN_CLASIFICAR: 0, TipoComentario.SIN_COMENTARIO: 0}
        resumen = analizador.resumen_comentarios()
        if salida_esperada == resumen:
            print("  ✔️ Resumen de comentarios generado correctamente.\n")
            for campo, votos in resumen.items():
                print(f"   {campo.name:>14}: {str(votos):<8}")
            print()
        else: raise(Exception)
    except Exception as e:
        print(f"  ❌ Error al generar resumen de votos: {e}\n")

    print("7. Acceso a propiedades:")
    try:
        encuestas = analizador.encuestas
        print("  ✔️ Acceso correcto a propiedades.\n")
        for encuesta in encuestas:
            print(encuesta,'\n')
    except Exception as e:
        print(f"  ❌ Error al acceder a propiedades: {e}\n")

    print("8. Intento de modificación de propiedades:")
    try:
        analizador.encuestas = {}
        print("  ❌ No se lanzó excepción al modificar 'encuestas'.\n")
    except AttributeError as e:
        print(f"  ✔️ No se puede modificar la propiedad 'encuestas'.\n")


# Ejecutar las pruebas
probar_analizador_encuestas()

=== PRUEBAS: AnalizadorEncuestas ===

1. Creación de AnalizadorEncuestas vacío:
  ✔️ Analizador creado con 0 encuestas.


2. Agregar encuestas:
  ✔️ Se agregaron 2 encuestas.


3. Promedio por campo:
  ✔️ Promedio por campo calculado correctamente: (4.0, 3.0, 3.5, 3.5, 2.5)

4. Promedio general:
  ❌ Error al calcular promedio general: 'AnalizadorEncuestas' object has no attribute 'promedio_general'

5. Resumen de votos:
  ❌ Error al generar resumen de votos: 'AnalizadorEncuestas' object has no attribute 'resumen_votos'

6. Resumen de comentarios:
  ❌ Error al generar resumen de votos: 'AnalizadorEncuestas' object has no attribute 'resumen_comentarios'

7. Acceso a propiedades:
  ✔️ Acceso correcto a propiedades.

Encuesta de cliente 'ABCD1234':
  Respuestas → atención: 5, calidad: 4, precio: 3, tiempo de entrega: 4, facilidad de uso: 5
  Comentario → [POSITIVO] Muy bien 

Encuesta de cliente 'WXYZ5678':
  Respuestas → atención: 3, calidad: 2, precio: 4, tiempo de entrega: 3, facilidad 

__Salida esperada__:
<pre style='font-size:12px;color:blue'>
=== PRUEBAS: AnalizadorEncuestas ===

1. Creación de AnalizadorEncuestas vacío:
  ✔️ Analizador creado con 0 encuestas.

	 No hay encuestas registradas. 

2. Agregar encuestas:
  ✔️ Se agregaron 2 encuestas.

Analizador de 2 encuesta(s):

--- Encuesta #1 ---
Encuesta de cliente 'ABCD1234':
  Respuestas → atención: 5, calidad: 4, precio: 3, tiempo de entrega: 4, facilidad de uso: 5
  Comentario → [POSITIVO] Muy bien

--- Encuesta #2 ---
Encuesta de cliente 'WXYZ5678':
  Respuestas → atención: 3, calidad: 2, precio: 4, tiempo de entrega: 3, facilidad de uso: 0
  Comentario → [NEGATIVO] Lento y caro
 

3. Promedio por campo:
  ✔️ Promedio por campo calculado correctamente: (4.0, 3.0, 3.5, 3.5, 2.5)

4. Promedio general:
  ✔️ Promedio general calculado correctamente: 3.30

5. Resumen de votos:
  ✔️ Resumen de votos generado correctamente.

         ATENCION: [5, 3]  
          CALIDAD: [4, 2]  
           PRECIO: [3, 4]  
   TIEMPO_ENTREGA: [4, 3]  
    FACILIDAD_USO: [5, 0]  

6. Resumen de comentarios:
  ✔️ Resumen de comentarios generado correctamente.

         POSITIVO: 1       
         NEGATIVO: 1       
           NEUTRO: 0       
      INAPROPIADO: 0       
   SIN_CLASIFICAR: 0       
   SIN_COMENTARIO: 0       

7. Acceso a propiedades:
  ✔️ Acceso correcto a propiedades.

Encuesta de cliente 'ABCD1234':
  Respuestas → atención: 5, calidad: 4, precio: 3, tiempo de entrega: 4, facilidad de uso: 5
  Comentario → [POSITIVO] Muy bien 

Encuesta de cliente 'WXYZ5678':
  Respuestas → atención: 3, calidad: 2, precio: 4, tiempo de entrega: 3, facilidad de uso: 0
  Comentario → [NEGATIVO] Lento y caro 

8. Intento de modificación de propiedades:
  ✔️ No se puede modificar la propiedad 'encuestas'. 
</pre>